File: examples/java-repo/src/test/java/com/example/ecommerce/domain/payment/PaymentTest.java
package com.example.ecommerce.domain.payment;

import static org.junit.Assert.*;

import com.example.ecommerce.domain.order.Money;
import org.junit.Test;

public class PaymentTest {

    @Test
    public void shouldCreatePendingPayment() {
        Payment payment = new Payment(1, 100, Money.of(50.00, "USD"), "credit_card");
        assertEquals(1, payment.getId());
        assertEquals(100, payment.getOrderId());
        assertEquals(PaymentStatus.PENDING, payment.getStatus());
        assertEquals("credit_card", payment.getMethod());
        assertNotNull(payment.getCreatedAt());
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldRejectZeroAmount() {
        new Payment(1, 100, Money.zero("USD"), "credit_card");
    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldRejectBlankMethod() {
        new Payment(1, 100, Money.of(50.00, "USD"), "");
    }

    @Test
    public void shouldProcessAndComplete() {
        Payment payment = new Payment(1, 100, Money.of(50.00, "USD"), "credit_card");
        payment.process();
        assertEquals(PaymentStatus.PROCESSING, payment.getStatus());
        payment.complete();
        assertEquals(PaymentStatus.COMPLETED, payment.getStatus());
    }

    @Test
    public void shouldProcessAndFail() {
        Payment payment = new Payment(1, 100, Money.of(50.00, "USD"), "credit_card");
        payment.process();
        payment.fail();
        assertEquals(PaymentStatus.FAILED, payment.getStatus());
    }

    @Test
    public void shouldRefundCompletedPayment() {
        Payment payment = new Payment(1, 100, Money.of(50.00, "USD"), "credit_card");
        payment.process();
        payment.complete();
        payment.refund();
        assertEquals(PaymentStatus.REFUNDED, payment.getStatus());
    }

    @Test(expected = IllegalStateException.class)
    public void shouldNotProcessNonPendingPayment() {
        Payment payment = new Payment(1, 100, Money.of(50.00, "USD"), "credit_card");
        payment.process();
        payment.process();
    }

    @Test(expected = IllegalStateException.class)
    public void shouldNotRefundNonCompletedPayment() {
        Payment payment = new Payment(1, 100, Money.of(50.00, "USD"), "credit_card");
        payment.refund();
    }
}

File: examples/go-repo/internal/domain/payment/payment_test.go
package payment_test

import (
	"testing"

	"github.com/example/go-repo/internal/domain/order"
	"github.com/example/go-repo/internal/domain/payment"
)

func TestNewPayment(t *testing.T) {
	amount, _ := order.NewMoney(50.00, "USD")
	p, err := payment.NewPayment(1, 100, amount, "credit_card")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if p.ID() != 1 {
		t.Errorf("expected ID 1, got %d", p.ID())
	}
	if p.Status() != payment.StatusPending {
		t.Errorf("expected pending, got %s", p.Status())
	}
}

func TestRejectZeroAmount(t *testing.T) {
	zero := order.ZeroMoney("USD")
	_, err := payment.NewPayment(1, 100, zero, "credit_card")
	if err == nil {
		t.Fatal("expected error for zero amount")
	}
}

func TestRejectEmptyMethod(t *testing.T) {
	amount, _ := order.NewMoney(50.00, "USD")
	_, err := payment.NewPayment(1, 100, amount, "")
	if err == nil {
		t.Fatal("expected error for empty method")
	}
}

func TestProcessAndComplete(t *testing.T) {
	amount, _ := order.NewMoney(50.00, "USD")
	p, _ := payment.NewPayment(1, 100, amount, "credit_card")
	if err := p.Process(); err != nil {
		t.Fatalf("process: %v", err)
	}
	if p.Status() != payment.StatusProcessing {
		t.Errorf("expected processing, got %s", p.Status())
	}
	if err := p.Complete(); err != nil {
		t.Fatalf("complete: %v", err)
	}
	if p.Status() != payment.StatusCompleted {
		t.Errorf("expected completed, got %s", p.Status())
	}
}

func TestProcessAndFail(t *testing.T) {
	amount, _ := order.NewMoney(50.00, "USD")
	p, _ := payment.NewPayment(1, 100, amount, "credit_card")
	p.Process()
	p.Fail()
	if p.Status() != payment.StatusFailed {
		t.Errorf("expected failed, got %s", p.Status())
	}
}

func TestRefundCompleted(t *testing.T) {
	amount, _ := order.NewMoney(50.00, "USD")
	p, _ := payment.NewPayment(1, 100, amount, "credit_card")
	p.Process()
	p.Complete()
	if err := p.Refund(); err != nil {
		t.Fatalf("refund: %v", err)
	}
	if p.Status() != payment.StatusRefunded {
		t.Errorf("expected refunded, got %s", p.Status())
	}
}

func TestCannotProcessNonPending(t *testing.T) {
	amount, _ := order.NewMoney(50.00, "USD")
	p, _ := payment.NewPayment(1, 100, amount, "credit_card")
	p.Process()
	if err := p.Process(); err == nil {
		t.Fatal("expected error processing non-pending payment")
	}
}

func TestCannotRefundNonCompleted(t *testing.T) {
	amount, _ := order.NewMoney(50.00, "USD")
	p, _ := payment.NewPayment(1, 100, amount, "credit_card")
	if err := p.Refund(); err == nil {
		t.Fatal("expected error refunding non-completed payment")
	}
}
